Block Editor: Simplify Selection Reducer

ellatrix opened the pull request:
## Description

Splitting out some stuff from #16428.

I've moved `isSelectionEnabled`, `isMultiSelecting`, and `initialPosition` out of the `blockSelection`, as these are not tied to the selection state. These can all be independent reducers, which is easier to maintain. Since the selection reducer didn't contain anything but the start and end of the selection, I flattened the structure to `state.selectionStart` and `state.selectionEnd`. The reducers ended up being much simpler.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

ellatrix committed:
Block Editor: Simplify Selection Reducer

mcsf commented on packages/block-editor/src/store/test/reducer.js:
```diff
@@ -1786,196 +1790,150 @@ describe( 'state', () => {
-			const original = deepFreeze( { start: { clientId: 'ribs' }, end: { clientId: 'ribs' }, isMultiSelecting: true } );
-			const state = blockSelection( original, {
-				type: 'STOP_MULTI_SELECT',
+	describe( 'selection()', () => {
```
Conceptually, should we be testing what is an internal helper for actually `selectionStart` and `selectionEnd`, or should we instead aim those tests at the final reducers? Strictly speaking, by testing `selection` we make no guarantees that `selection{Start,End}` follow that same behaviour.

ellatrix replied:
Yes, that's true. We can test `selectionStart` and `selectionEnd` instead...

ellatrix replied:
I changed the tests.

mcsf commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -922,161 +922,177 @@ export function isCaretWithinFormattedText( state = false, action ) {
-
 /**
- * Reducer returning the block selection's state.
+ * General reducer returning the block selection's start and end.
```
Perhaps just to me, but _general reducer_ isn't enough to convey that this is only an internal helper for two other reducers. Can we make it more explicit? Maybe:

> Internal helper reducer for `selectionStart` and `selectionEnd`. Can hold a block selection, represented by an object with property `clientId`.

mcsf commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -922,161 +922,177 @@ export function isCaretWithinFormattedText( state = false, action ) {
+	} else if ( action.type === 'REMOVE_BLOCKS' ) {
+		return state;
+	}
+}
```
This reducer is a little unconventional, in that its default behaviour isn't to return `state`, but rather to implicitly return `undefined`. Is this the best way to proceed? Seems like it could increase the likelihood of mistakes in the future.

ellatrix replied:
How so? The reducer should store the initial position temporarily, until another action occurs. The only exception is `REMOVE_BLOCKS`. As mentioned above in the doc, the reducer should eventually be removed in favour of setting selection directly.

mcsf replied:
This is more or less the same observation as in https://github.com/WordPress/gutenberg/pull/17482#discussion_r328165271

I'll defer to your judgment, as I don't want to block this.

ellatrix replied:
Ok, I'll add a comment.

mcsf commented on packages/block-editor/src/store/test/reducer.js:
```diff
@@ -22,7 +22,11 @@ import {
+	selectionStart,
+	selectionEnd,
+	initialPosition,
+	isMultiSelecting,
```
These flattened reducers have some overlap in the types of actions to which they react. One thing that worries me then is that, by only testing them as independent reducers, we may be omitting useful tests that look at the compound state of these reducers.

Could we keep this state compound while still benefitting from the simplicity of flattened reducers? What if, in `reducers`, we still write `selectionStart` _et al._ as independent functions, but at the end combine them as `blockSelection: combineReducers( { selectionStart, selectionEnd, initialPosition, isMultiSelecting } )`? Or is this just completely against the objectives behind this PR?

ellatrix replied:
What's the overlay? `selection` and `selectionStart`/`selectionEnd` don't have any overlap. `isMultiSelecting` doesn't have any overlap with the others either. `initialPosition` has a very limited overlap, where the state is very different from the rest. I don't really see a problem here.

mcsf reviewed (commented):
Overall this looks like a step in the right direction, but I left a couple of questions.

ellatrix committed:
Test selectionStart and selectionEnd reducers instead of internal selection reducer

mcsf reviewed (approved):
If you're confident about test coverage as reducers become decoupled, I trust you!

ellatrix committed:
Add comment when reducer resets state by default

ellatrix commented:
Thanks @mcsf! I adjusted the tests so both the final `selectionStart` and `selectionEnd` are tested. I also added a comment for the unconventional reducer.
